### ISSUE TITLE
Discussion - Acro flight performance enhancements

### ISSUE THREAD

**Issue by theArchLadder** [iNavFlight/inav#218]
I just tried betaflight, wow! Very impressive, so locked in compared to fp-pid. It felt like i became a twice as good pilot!

Could we get this performance in inav? Why having to compromise between nav and acro performance! :stuck_out_tongue_winking_eye: 
The default looptime on betaflight is 500 and even i could notice a difference when switching to 1000 like i have on inav. 500 made it feel a little bit more sharp.
Would it be possible to get inav to run at 500 on F3 targets?

But the most important difference is that the attitude was so locked in!
Perhaps i'm a terrible tuner but no matter what i do i can't get the attitude to be locked in on fp-pid, and to high I-gains starts to ruin the performance... Is there to much anti-windup or what is going on here? Attitude on betaflight was very locked in even on 1000 looptime. Fp-pid changes the attitude all the time without me telling it to, that makes acro flying very hard.

As usual i'm more than happy to make my custom builds and flight test them, just give me some ideas on what to try! What else could we borrow from betaflight to improve acro flight?

**Comment by DzikuVx** [comment]
If you want to have much better performance for acro:
- set gyro_lpf=188HZ
- set gyro_soft_lpf_hz=100
- set dterm_lpf_hz=50
- set gyro_sync=ON
- set gyro_sync_denom=1

**Comment by theArchLadder** [comment]
@DzikuVx Thanks! :smile: 
I will do a test flight tomorrow!

My biggest problem is the changing attitude though. For example when i go fast from side-to-side using only roll, pitch and yaw will change attitude, but this does not really happen in betaflight. 
It would be interesting to hear what you, or @digitalentity think the reason is for the unstable attitude and what can be done about it...?

It would also be helpful if i could convert the defaults of betaflight mwrewrite into fp-pid, does anyone know how that can be calculated?

**Comment by DzikuVx** [comment]
For attitude.... More I? I have around 60 on roll and 90 on pitch.

**Comment by DzikuVx** [comment]
42hz for gyro is noise proof but barely flyable. 98 is compromise. 188 is afair what betaflight uses. 
And what are your P?

**Comment by theArchLadder** [comment]
gyro_lpf = 98HZ
set p_roll = 35
set i_roll = 80
set d_roll = 80

My P was 50 when my quad was 150g heavier. AUW now is 500g, with 6 inch props and 4s battery.
I did try I at 100 but i started getting bounce-back and it just didn't feel right.

I plan to try cleanflight tomorrow to see how that feels...

**Comment by digitalentity** [comment]
Lower looptimes tend to tolerate vibrations better. You might want to
adjust your filtering settings on D and gyro.

Blackbox logs will help debug. Possibly too low P and too high D, hard to
guess.

We can't have looptime faster than 1000 at the moment - requires changes to
scheduler code and decoupling of gyro, accelerometer reads, PID controller
and motor updates.

Will work in that direction after configuration refactoring PR is merged.

**Comment by digitalentity** [comment]
Can you post compete dump here?

**Comment by theArchLadder** [comment]
This blackbox log should be decent, i was flying fast side-to-side using only roll, that's when i had the issues with the attitude wandering about.
https://www.dropbox.com/s/4yog9m5wd8mht51/blackbox_log_2016-05-17_163714_pedestrian-limit-test?dl=0

Dump:
https://www.dropbox.com/s/bk9fovvj8n7gd6t/2016-05-16_x3_theladder?dl=0

Is it possible to convert between fp-pids and the pids of mwrewrite or luxfloat? That way i can compare with the same pid-tuning.

**Comment by theArchLadder** [comment]
I just tried cleanflight 12.0, and it also felt very good, the main difference is that the attitude doesn't drift. When doing quick roll manuvers, pitch and yaw stays the same. When doing yaw pirouttes, pitch/roll stays the same, this is not my experince with inav...

I need to try inav again now and see how big the difference really is, and do some more tuning, but something feels wrong here, cleanflight with default settings should not be better than inav for acro, right?

**Comment by Ralf-W** [comment]
@theArchLadder Just some observations in your BB log. At the end of a roll command - roll (and pitch) stick are centered - we see the roll gyro and the rolling copter overshooting several times with reducing amplitude. Some gain is too high.
Before the roll you made the copter climb with throttle. Then throttle is reduced to get a roll with low hight loss. However, throttle gets so low, that two motors are more or less stopping during the roll. Looking at the 4 motor graphs, the rpm are not controlled symetrically. Two curves have a sharp Peak. The lower ones get cut in a plateau and cannot be reduced as far as the other two are incresed.
![roll_1](https://cloud.githubusercontent.com/assets/16436064/15447207/9fb6c60c-1f39-11e6-8a40-07abff07b013.jpg)
Although there is no rc command, pitch and yaw come alive due to the asymmetrical rpm changes of the 4 motors.
![roll_2](https://cloud.githubusercontent.com/assets/16436064/15447258/b877995c-1f3c-11e6-86d2-97ee1d4db22f.jpg)

In a roll with high throttle setting, two motor rpm are maxed out. This results in an oscillation in the pitch and even in the yaw directions (yellow gyro curve is yaw, turquise gyro curve is pitch).
![roll_high_throttle](https://cloud.githubusercontent.com/assets/16436064/15447302/60b61d72-1f3e-11e6-9a99-9efae5afd016.jpg)

**Comment by theArchLadder** [comment]
@Ralf-W Thanks! Do you have any idea what's wrong with my pids? To me it looks like i-gain might be to high but i'm not that good at pid tuning

My blackbox denom is way to high in that log, problem is though, if i have it at 1-2 my black box fills up super quick... I guess i can do som super short flights and look at the blackbox.

**Comment by Ralf-W** [comment]
@theArchLadder I updated my comment. Please have a second look at it.

I am still no expert in tuning, but I would start reducing the I-gain, too. And even the D-gain can be set to Zero, just to make analysis easier.

**Comment by theArchLadder** [comment]
@Ralf-W Oh wow, thank you very much for that detailed analysis!

I think i figured out how to change the pid scaling to match betaflight. I will try betaflight defaults and see what happens...

**Comment by theArchLadder** [comment]
Hmm, tried some more betaflight and this time tuned only p so that it barely had any oscillations, default p=45 and i ended up with p=20, and it actually seemed to fly well so perhaps i have been using way to high pids... I guess it makes sense since my quad is lighter than the average fpv racer, and has 6 inch props.

Still not sure why both cleanflight and betaflight seems to hold attitude better than inav during agressive acro flying, when i use stock pids on all of them...

EDIT: Did some tuning with inav, feels different from betaflight while tuning, oscillations are slower and i find it harder to tune to fly well. Time to play with some lowpass filters i guess

EDIT2: Used these settings
gyro_lpf = 188HZ
gyro_soft_lpf_hz = 100
dterm_lpf_hz = 100
I did not notice any improvement. On betaflight i use 50 for both gyro_soft and dterm, i guess betaflight does not use the gyro hardware filter anymore...

EDIT3: same 1000 looptime was used on all tests. I wish inav still had luxfloat so i could compare...

**Comment by theArchLadder** [comment]
Ugh, it's a nightmare debugging this. Both cleanflight and betaflight flies good on defaults, but inav is unstable just looking at the quad during a hover, and has problem keeping attitude etc... And i have not manage to tune it to behave good

It would really help to have the same pid scaling. Both luxfloat and mwrewrite has the same scaling in both betaflight and cleanflight, so it would be helpful if inav used the same scaling.

I tried to use this:
https://github.com/cleanflight/cleanflight/blob/52612ebb3a9ae4668a021a51c75de1c726977e1a/src/main/flight/pid_luxfloat.c#L70
In inav, and ended up replacing these:

```
#define FP_PID_RATE_P_MULTIPLIER  31.219513f    // inav 40.0f    // betaflight - 40.0
#define FP_PID_RATE_I_MULTIPLIER  4.092004f     // inav 10.0f    // betaflight - 10.0
#define FP_PID_RATE_D_MULTIPLIER  1890.629883f  // inav 4000.0f  // betaflight - 1000.0
```

Those betaflight comments are already in inav, but i think those are from before betaflight changed luxfloat scaling to match mwrewrite.
I tried these scalings with cleanflight default pids, but i got lots of oscillations that i did not have in cleanflight, so perhaps i did something wrong. I think i need @digitalentity to help me get this scaling right...

Oh, i also tried to completely disable anti-windup but that didn't improve things, nothing wrong with those it seems, good :+1:

**Comment by oleost** [comment]
@theArchLadder Try this: https://docs.google.com/spreadsheets/d/1mX0a6yytARsZiaU6sCGKvpwXSBYjTjTwVJZiJmktKqg/edit?usp=sharing

This is atleast what DigitalEntity told me to convert from old luxfloat to new FP PID

EDIT: Use values from CLI

**Comment by oleost** [comment]
@theArchLadder And remember use values from CLI.

**Comment by theArchLadder** [comment]
@oleost Thanks... Are those luxfloat values the old defaults? In that case this seems obsolete, fp-pid p-gain of 200 is very high compared to default 30...

**Comment by oleost** [comment]
No, those are my PIDs that I had on my fixed wing. So yes, they are very high and unusable on MR

**Comment by oleost** [comment]
@theArchLadder So you got it to fly better?

Still looks like  people are having more issues with iNav and default values than with cleanflight and betaflight.

**Comment by DzikuVx** [comment]
I took some paper, a pen and did some calculations today. I will share my thoughts later in a spearate ticket, but looks like it is as issue of not only default PIDs but also scaling (how response scales to PID)

**Comment by theArchLadder** [comment]
Awesome that we have the same pid scaling as betaflight/cleanflight now! I have been flying betaflight and switched over to inav just now with the same pids, and my first impression is that it does not feel as locked in.

Settings used:
- set gyro_lpf=188HZ
- set gyro_soft_lpf_hz=75
- set dterm_lpf_hz=50
- set gyro_sync=ON
- set gyro_sync_denom=1

@digitalentity @DzikuVx 
Does inav use the same gyro software filter as betaflight?
AFAIK betaflight turns off the hardware gyro filter, but this does not seem possible in inav, could we bring that over?

**Comment by digitalentity** [comment]
@theArchLadder can you post your dump from betaflight?

**Comment by theArchLadder** [comment]
@digitalentity I didn't save that, but almost everything was on defaults, except gyro_lowpass=75. Is there anything specific you want to know?

I did some more flying and it behaved better when i increased pitch/roll d-gain to 35, i have used 18-25 on betaflight. Perhaps inav need higher d-gain because of this? https://github.com/iNavFlight/inav/blob/master/src/main/flight/pid.c#L274

**Comment by DzikuVx** [comment]
There is a very specific reason why Betaflight can be so good in terms of Acro: type of aircraft it is targeted to run on: 250 or smaller with a very rigid frame using 5" or smaller props. Weighting usually between 500 and 700g. And thanks to that:
- small and rigid frame will rather attenuate motor noise instead of increasing it because of resonnance, harmonics and things like this
- motors usually operate on >8000RPM and that means they generate >133Hz noise
- in terms of flight frequencies > 100Hz are useless, so it's super easy to cut all the noise and leave what we need.  Just set cutoff at half of hover noise frequency and you are happy to go (huge simplification)
- things like looptime are only secondary
- lower cutoff = bigger delay, bigger error to correct when signal gets in, more problems during tuning

With bigger machines it's not so easy. For example, my 600mm on 12" props quad hovers at about 4500RPM. Noise it then around 75Hz. Cutoff of 75Hz would make it unflyable: all the motor noise would go straight into gyro.  Even for 10" props cutoff 75Hz is too high. 

INAV is not targeted at any kind of aircraft. Everyone expects his machine would fly great at default values. Pretty impossible. We can make it flyable for everyone. But not flying good for everyone. 

There is only one  solution I see: presets in configurator. I've already put so thinking into it, and this is how I see it: new configurator tab when pilot can choose propeller size and type: racer, photo platform, whatever. Maybe also weight, frame size... but probably only propeller size would be really enough. Preset would set gyro, LPF cutoffs, maybe something else too. Of maybe even some PID changes.

I think I will start working on this somewhere after my vacation.

**Comment by martinbudden** [comment]
@theArchLadder , have you also adjusted your rates? Your rates will have an effect on the feel and they are scaled differently in iNav and betaflight, see https://github.com/iNavFlight/inav/wiki/Rate-conversion-for-INAV-1.2

@DzikuVx , I think having presets in the configurator would be excellent. Openpilot/Librepilot has something similar (at least I assume it does - as part of the setup wizard it asks you to choose a frame type and I presume that is used to set PIDs, filters etc - I haven't actually looked at the code though). And Openpilot flies very well "out of the box" - ie with no tuning.

**Comment by DzikuVx** [comment]
@martinbudden Exactly, OpenPilot was source of my idea. I also not sure what was set in those, but PIDs and rates for sure.

**Comment by theArchLadder** [comment]
@DzikuVx I'm okay with having to tune filters and pids on inav, but so far i have not been able to come close to default betaflight. Inav feels sloppy and lose while betaflight feels locked in.

The main difference between bf and inav afaik is that bf can turn off the hardware gyro filter, as long as the software filters are the same in bf and inav (are they?)

@martinbudden Yes, i use 540deg/s for now, feels about right to me

**Comment by martinbudden** [comment]
@theArchLadder , software filtering of DTerm is different in iNav and bf. iNav uses a filter to differentiate and filter at the same time.

**Comment by theArchLadder** [comment]
Wow, @DzikuVx was not lying, filters make a huge difference. I did an experiment like this:
i-gain and d-gain at zero and looptime at 1000, then do a gentle hover and increase p-gain until i got constant oscillations

**Betaflight**

| p-gain | software gyro lpf | hardware gyro lpf |
| --- | --- | --- |
| 8.7 | 100 | off |
| 8.4 | 75 | off |
| 6.7 | 100 | 188hz |
| 6.6 | 75 | 188hz |

**inav**

| p-gain | software gyro lpf | hardware gyro lpf |
| --- | --- | --- |
| 5.8 | 75 | 188hz |
| 4.4 | 75 | 42hz |

As you can see, turning off the hardware lpf allows much higher p-gain, i get why betaflight has it turned off by default now. I guess i should test different software lpf values without hardware lpf and see how that performs

**Comment by DzikuVx** [comment]
With lpf problem is that it is also platform dependent. If you have a lot of noise, you need lower cutoff to increase atenuation. If you have super super smooth motors and props, you can increase cutoff. 
Disabling hardware lpf means, in theory, you should lower software cutoff to compendate for missing 3dB. But lpf is not libear. 
Plus of having only one lpf is lower delay. 

Ah, are you sure gyro_lpf 256 is not in deed the same as OFF?

**Comment by theArchLadder** [comment]
Did some more tests on betaflight:

| p-gain | software gyro lpf | hardware gyro lpf |
| --- | --- | --- |
| 6.4 | 50 | off |
| 4.1 | 25 | off |
| 10.5 | 200 | off |

the results with low p-gain has slow and big oscillations, while high p-gain gives small and fast oscillations. Seems like a good way to test how much delay there is in the system.

I have not compared how effective soft vs hard is at removing noise, but i know that only soft seems to be effective on betaflight. 188hz hardware lpf sounds to high to be useful but it still seems to add lots of delay. The betaflight guys seems to know what they are doing and it seems like they came to the conclusion that using only software lpf is better.

@DzikuVx not sure if 256hz does it, IIRC i was unable to arm with that last time i tried, i guess i have to reflash inav and try it.

**Comment by theArchLadder** [comment]
@DzikuVx Something seems wrong with 256hz option, after i set it the inav configurator just says "waiting for data" no matter what tab i'm in, even after restarting both configurator and FC...

Oh, i just figured it out, i have to change `set gyro_sync_denom=8` to get 1000 looptime, i guess the gyro starts to run in 8khz mode.

**Comment by theArchLadder** [comment]
Yay it worked! Thanks @DzikuVx !
- gyro_lpf = 256HZ
- gyro_sync_denom=8
- gyro_soft_lpf_hz = 100

Those settings gives constant oscillations at around 8.5 p-gain, just like betaflight, nice! Feels like we just got the P-term to behave as good/responsive as betaflight, i guess the next thing to check is D-term and see if there is any more delay there compared to betaflight, i don't think i will have time for that today though.

Inav uses a fancy "5-point noise-robust differentiators without time delay":
https://github.com/iNavFlight/inav/blob/master/src/main/flight/pid.c#L274
Perhaps this will allow a higher `dterm_lpf_hz` so we can reduce delay without getting more noise. Or perhaps that "without time delay" is a lie and we would be better of without it :stuck_out_tongue_winking_eye: 

EDIT: Looks like we reduced the gyro delay so i guess that should also help with d-term...

**Comment by DzikuVx** [comment]
Ha, this is a very nice think we learned here. 
INAV will never do looptimes as low as Betaflight because of higher computational requiements, but I see no reason not to be quite close

**Comment by theArchLadder** [comment]
Did some more test flights just now, i ended up with these settings:
gyro_lpf = 256HZ
gyro_sync_denom=8
gyro_soft_lpf_hz = 75
dterm_lpf_hz = 50
p_roll = 45
i_roll = 30
d_roll = 30
roll_rate = 72

There is still a bit of noise when looking in the bblog, the "5-point noise-robust differentiators without time delay" does not seem to make any big difference as far as i can tell. But i think the noise is on an acceptable level.

But wow, now it actually starts to fly well! One good sign of this is that my quad used to be very unstable and wobble a lot during -200cm/s descents, but now it looks really stable! Awesome! I can sort of hear the motors working hard to fight the turbulent propwash but the quad still looks stable, big improvement, YAY! :smiley:

**Comment by digitalentity** [comment]
@theArchLadder 
`gyro_lpf = 256HZ` is actually no LPF on MPU end, so this is OFF setting.

Ok, so this issue is basically a matter of properly setting parameters for miniquads. Thanks for figuring this out!

**Comment by theArchLadder** [comment]
Not sure if this is only for miniquads though, great acro gives great angle performance, and that should help with navigation stuff, surly bigger copters also want a wobble-free descent for example. Perhaps turning off the hardware lpf (also using 8khz mode) and using a lower software lpf hz will reduce delay but still be effective at reducing noise, and therefore increase performance...?

Betaflight also seems to do some clever stuff with i-term, i think that's worth investigating too.

**Comment by DzikuVx** [comment]
I took my 250 for a spin with

```
gyro_lpf = 256HZ
gyro_sync_denom=12
gyro_soft_lpf_hz = 75
dterm_lpf_hz = 50
```

and I must say, that was pretty impressive. Maybe not super awesome because of one motor slightly off balance, but pretty impressive. Specially prop-wash situations improved greatly. Also rolls seems to be much more dynamic now.

**Comment by martinbudden** [comment]
@DzikuVx , @digitalentity , I think as well as @DzikuVx idea for the configurator setting default values, I think we should revisit the default values in iNav. In particular I favour less hardware filtering, and compensating with different soft filters.

@DzikuVx , it will also be interesting to hear any results you have with your larger quads when you get back from holiday.

(My largest quad is a 250, but I'm looking at building a 350 with camera and gimbal soon...)

**Comment by DzikuVx** [comment]
@martinbudden from my latest experiments I think that only software LPF might be a good idea also for big machines. Before one of my motors started to introduce much noise I was using `gyro_lpf=188HZ` and  `gyro_soft_lpf_hz=50`. I will try to confirm that as soon as I replace faulty motor

**Comment by digitalentity** [comment]
As a matter of fact on a 450-sized machine without Damped Light support on the ESCs I was only able to get a flyable tune with gyro_lpf=98HZ or 42HZ. Anything higher made D-term crazy no matter how low I was setting gyro_soft_lpf_hz.

I guess heavy internal filtering with gyro_lpf introduces a lot of delay which makes closed-loop control worse on all machines, however on bigger machines high-frequency vibrations might make things even worse.

The trick is that MPU is sampling gyro internally at 8kHz and thus is able to do very effective LPF filtering. If we would be sampling gyro at 8kHz as well, we could always be running with imternal LPF disabled.

As a matter of fact I'm thinking about modifying gyro driver to read gyro at highest possible rate (1kHz or 8kHz depending on gyro_lpf setting) and downsample it to looptime if required. Current gyro_sync_denom setting is dropping unused samples, I intend to average them to suppress noise.

**Comment by digitalentity** [comment]
For example:

```
gyro_sync=256HZ
gyro_sync_denom=16
```

will yield 500Hz looptime and will use only 1 sample out of 16 (the rest 15 would be silently dropped). If we modify the code to record all 16 samples and average them out we will significantly suppress influence of high-frequency noise and will reduce effect of aliasing (current approach will alias everything above 250Hz, proposed approach will be immune to aliasing up to 4kHz vibrations).

**Comment by digitalentity** [comment]
I guess it's time to look into making INAV better suited for tiny acro machines - it flies good enough on bigger ones :smiley:

**Comment by martinbudden** [comment]
@digitalentity, as you point out, setting `gyro_sync_denom` to a value of other than one results in gyro samples being dropped rather than used.

Before I got diverted by the F4 port I was working on this issue. However, rather than change the gyro driver, I think the correct approach is to make changes to the PID controller.

My ideas were actually prompted by the way you have structured the iNav PID. As stated in your code, it is divided into 4 steps:

```
// Step 1: Calculate gyro rates
// Step 2: Read target
// Step 3: Run control for ANGLE_MODE, HORIZON_MODE, and HEADING_LOCK
// Step 4: Run gyro-driven control
```

However, all 4 steps are called sequentially from within the main `pidController` function.

My code, which I've provisionally implemented in a new PID (which I've called "Mars") at https://github.com/martinbudden/cleanflight/blob/pid_mars/src/main/flight/pid_mars.c instead spits the PID into 3:
1. `pidMarsUpdateGyroRate` - called every time there is a `gyroUpdate`, regardless of `gyro_sync_denom`
2. `pidMarsUpdateDesiredRate` - called whenever there is new RC input (ie the pilot has changed the desired rate)
3. `pidMarsCalculate` - called whenever a calculation is required - ie just before the motors are updated

This means each part of the pid calculation is only performed when it is required and so `pidMarsUpdateGyroRate` can be called at a high frequency.

`pidMarsUpdateGyroRate` does three things:
1. updates a filter using the latest gyro rate. Currently I'm using a FIR filter, which I think gives better results than simply averaging the gyro value
2. updates ITerm (ie adds the gyro rate and does some anti-windup calculations)
3. optionally applies some DTerm filtering

The advantages of this approach are
1. No gyro update values are dropped
2. Specific filtering actions can be taken for P, I and D terms, rather than having generic filtering or averaging in the gyro driver.
3. Since calculations are only peformed when required, overall overhead of PID is reduced
4. `taskMainPidLoopChecker` in `mw.c` can be simplified (although in my code it is more complicated, since I support both the old PIDs and the new).

I sort of regard this as the logical conclusion of the restructuring work you did for the iNav PID and think it will result in much improved performance, although I have only done very limited testing.

**Comment by theArchLadder** [comment]
At what rate can the gyro be read when connected with i2c?

**Comment by digitalentity** [comment]
@theArchLadder 
If we would have a proper DMA-based I2C driver, we can reach 3-4kHz gyro rate even with I2C bus at 400kHz. Problem is that currect I2C driver is blocking meaning that going over 1-2kHz would be problematic already.

**Comment by digitalentity** [comment]
@martinbudden 
I'd separate gyro completely from PID loop. Read gyro as fast as possible, filter it and provide value to PID loop only when necessary. Make gyro and PID loop run at different rate (possibly even asynchronous).

**Comment by gurkenfolie** [comment]
tried  my 250 with this settings
gyro_lpf = 256HZ
gyro_sync_denom=12
gyro_soft_lpf_hz = 75
dterm_lpf_hz = 50
F1, looptime 2000

the quad flies way better, rolls are very smooth, poshold behaviour is better, plus i was able to go up with P up to 16 without oscillations (felt so, i will have a look at the blackbox log).

**Comment by DzikuVx** [comment]
@digitalentity @martinbudden Idea of separating gyro readouts and PID loop sounds very promising to me.

**Comment by DzikuVx** [comment]
I'm trying to find an information how much delay BiQuad filter would induce for 8kHz mode. I've found delay information for MPU6050

**Comment by martinbudden** [comment]
@digitalentity , I think we are thinking along similar lines here, perhaps with some differences in the implementation details.

But my work on this is firmly on hold a the moment - need to get the F4 port and the parameter groups done first...

**Comment by digitalentity** [comment]
@martinbudden I'm putting all further improvements on hold until F4 and Param Groups are implemented, to prevent conflicts.

**Comment by DzikuVx** [comment]
@martinbudden @digitalentity @theArchLadder  after replacing faulty motor I did some experiments with `gyro_soft_lpf_hz` on my big machine: 12" props, 600mm, 1.8kg. And results are well, interesting. 

Looks like the same principle applies to small and big machines after all. Last entry from blackbox log shows what is happening at `gyro_soft_lpf_hz=80` and `gyro_lpf=188HZ`, `gyro_sync=ON` and `gyro_sync_denom=2`. OK, gyros and motors are quite jumpy and far from clean, but quad flies very well. Is responsive, holds attitude, descends with only small wobble. 

What is even stranger, current cutoff is _above_ motor noise at hover (around 75Hz). Looks like, gyro noise has less effect that signal delay. With `gyro_lpf=188HZ` and `gyro_soft_lpf_hz=50` or even `gyro_soft_lpf_hz=45` gyro traces looked better, but **_P**_ gains had to be much much higher (around x1.5 higher) to archive similar effects. But then quad was very machine-like. 

Probably huge inertia of 12" props acts as super efficient LPF filter that breaks feedback loop PID->motor->gyro->PID

```
P = 85
I = 16
D=33
```

My next conclusion is that high _I_ gain is needed only on acrobatic machines. On big machines, where stability is more important than flips or roll (who does flips on 2kg UAV?) low _I_ gain gives much better results.   

Also, in such non-acribatic machines D term has very little impact on performance. D term trace (not counting noise) is almost all the time flat.

[bb_krogulec_lpf_wysoko.TXT](https://github.com/iNavFlight/inav/files/346097/bb_krogulec_lpf_wysoko.TXT)

**Comment by DaleKramer** [comment]
This is a great discussion! I am about to start trying to tune my 6 kg sort of hexacopter  [http://www.vlazair.com/projectStatus.html](http://www.vlazair.com/projectStatus.html) (only 1.3/1 thrust to weight), any suggestions on where I should start with all those parameters you are playing with?

**Comment by DzikuVx** [comment]
@DaleKramer I would say start with defaults. I see you have some fancy design over there, so nobody knows how it's gonna behave. At least in vertical flight. 
About transition to horizontal... hmm... I really do not think INAV is prepared to do smth like this...

**Comment by DaleKramer** [comment]
Yep, but someone has to try ;).  I am just thinking 80 degree angle mode will let me transition to horizontal flight and a small aerodynamic zoom to high alpha and then start up the electric motors will work for transition back to hover.  I will use idle up to start/stop electric motors and FC.

**Comment by digitalentity** [comment]
@DaleKramer wow, that's a fancy design! You are very much welcome to bring VTOL capabilities to INAV :smiley: 
At the moment INAV doesn't have the capability to transition from multirotor mixer to airplane mixer as they use different servo mapping, but that's something we should account for. VTOLs are great!

**Comment by DaleKramer** [comment]
I am operating my aircraft servos in parallel with FC commands, using TREA123 stick mode so hovering uses same logic (frame of reference) as 3D hover in aircraft. FC will be off using idle up OFF in horizontal flight.

**Comment by digitalentity** [comment]
Hm... using simple airplane mixer might work. Make servos and motors work together in vertical flight using some smart mixing. Interesting.

**Comment by DaleKramer** [comment]
Hoping to start tuning pitch axis on a test stand very soon, time will tell.

**Comment by gurkenfolie** [comment]
i checked my bb log, D goes crazy with  gyro_lpf = 256HZ, when set to the default value everything is fine but P has to be lowered a lot. with gyro_lpf = 256HZ the quad flies better but the motors are somewhat nervous.

**Comment by theArchLadder** [comment]
@gurkenfolie You can lower these:
gyro_soft_lpf_hz = 75
dterm_lpf_hz = 50
But it sounds like you have to much vibrations, fixing that would be even better.

**Comment by gurkenfolie** [comment]
gyro_soft_lpf set to 60 helped.
i will give it a new try with new balanced propellers and one motor has to be changed, maybe this causes the vibrations.

**Comment by skaman82** [comment]
Just testet the latest master with my 250 quad with 6" props. Angle, Horizon and GPS modes a working really good but it is practically unflyable in acro. Tried to change some settings to

tried my 250 with this settings
`gyro_lpf = 256HZ`
`gyro_sync_denom=12`
`gyro_soft_lpf_hz = 75`
`dterm_lpf_hz = 50`
`F3, looptime 2000`

but it made things much worse.

**Comment by theArchLadder** [comment]
@skaman82 "worse" doesn't tell us much, in what way was it worse?
Did it not feel locked in? Did you have oscillations? Was the stick to fast/slow?

**Comment by skaman82** [comment]
Well no oscillations but it flies like a drunken sailor. Hard to describe The quad completely breaks out in fast maneuvers and yaw turns. It is very opposite of locked in.

**Comment by skaman82** [comment]
With changes i posted above it was absolutley not flyeble at all. Even in stabilizised modes. I barely could control it

**Comment by theArchLadder** [comment]
@skaman82 Sounds very strange, can you post a full dump please?

**Comment by Ralf-W** [comment]
Please post your complete CLI dump. Then we can see if all parameters fit. E.g. did you "set gyro_sync = ON"? On my Naze32  "gyro_sync_denom=12" is resulting in a looptime of 1500 but you post looptime 2000.

**Comment by skaman82** [comment]
Here is my dump and my blackbox log - but I'm sure it run out of flash space BEFORE I tested Acro and the custom settings I mentioned above:

# version

# INAV/SPRACINGF3 1.2.0 Jul 16 2016 / 21:18:21 ()

# pflags

# Persistent config flags: 0x00000001

# dump master

# mixer

mixer QUADX
mmix reset
smix reset

# feature

feature -RX_PPM
feature -VBAT
feature -INFLIGHT_ACC_CAL
feature -RX_SERIAL
feature -MOTOR_STOP
feature -SERVO_TILT
feature -SOFTSERIAL
feature -GPS
feature -FAILSAFE
feature -SONAR
feature -TELEMETRY
feature -CURRENT_METER
feature -3D
feature -RX_PARALLEL_PWM
feature -RX_MSP
feature -RSSI_ADC
feature -LED_STRIP
feature -DISPLAY
feature -ONESHOT125
feature -BLACKBOX
feature -CHANNEL_FORWARDING
feature -RX_NRF24
feature -SOFTSPI
feature RX_PPM
feature VBAT
feature MOTOR_STOP
feature GPS
feature FAILSAFE
feature LED_STRIP
feature BLACKBOX

# beeper

beeper GYRO_CALIBRATED
beeper RX_LOST
beeper RX_LOST_LANDING
beeper DISARMING
beeper ARMING
beeper ARMING_GPS_FIX
beeper BAT_CRIT_LOW
beeper BAT_LOW
beeper GPS_STATUS
beeper RX_SET
beeper ACC_CALIBRATION
beeper ACC_CALIBRATION_FAIL
beeper READY_BEEP
beeper MULTI_BEEPS
beeper DISARM_REPEAT
beeper ARMED
beeper SYSTEM_INIT
beeper ON_USB

# map

map AETR1234

# serial

serial 0 1 115200 57600 0 115200
serial 1 1 115200 57600 0 115200
serial 2 2 115200 115200 0 115200

# led

led 0 6,1::WC:6
led 1 6,0::C:6
led 2 7,0::C:6
led 3 8,0::C:6
led 4 9,0::C:6
led 5 10,0::C:6
led 6 10,14::WC:2
led 7 10,15::AC:2
led 8 9,15::C:2
led 9 8,15::C:2
led 10 7,15::C:2
led 11 6,15::AC:2
led 12 0,0:::0
led 13 0,0:::0
led 14 0,0:::0
led 15 0,0:::0
led 16 0,0:::0
led 17 0,0:::0
led 18 0,0:::0
led 19 0,0:::0
led 20 0,0:::0
led 21 0,0:::0
led 22 0,0:::0
led 23 0,0:::0
led 24 0,0:::0
led 25 0,0:::0
led 26 0,0:::0
led 27 0,0:::0
led 28 0,0:::0
led 29 0,0:::0
led 30 0,0:::0
led 31 0,0:::0

# color

color 0 0,0,0
color 1 0,255,255
color 2 0,0,255
color 3 30,0,255
color 4 60,0,255
color 5 90,0,255
color 6 120,0,255
color 7 150,0,255
color 8 180,0,255
color 9 210,0,255
color 10 240,0,255
color 11 270,0,255
color 12 300,0,255
color 13 330,0,255
color 14 0,0,0
color 15 0,0,0

set looptime = 2000
set emf_avoidance = OFF
set i2c_overclock = OFF
set gyro_sync = OFF
set gyro_sync_denom = 2
set mid_rc = 1500
set min_check = 1100
set max_check = 1900
set rssi_channel = 0
set rssi_scale = 30
set rssi_ppm_invert = OFF
set rc_smoothing = ON
set input_filtering_mode = OFF
set min_throttle = 1150
set max_throttle = 1850
set min_command = 1000
set servo_center_pulse = 1500
set 3d_deadband_low = 1406
set 3d_deadband_high = 1514
set 3d_neutral = 1460
set 3d_deadband_throttle = 1000
set motor_pwm_rate = 400
set servo_pwm_rate = 50
set disarm_kill_switch = ON
set auto_disarm_delay = 5
set small_angle = 35
set reboot_character = 82
set gps_provider = UBLOX
set gps_sbas_mode = NONE
set gps_dyn_model = AIR_1G
set gps_auto_config = ON
set gps_auto_baud = ON
set inav_auto_mag_decl = ON
set inav_accz_unarmedcal = ON
set inav_use_gps_velned = ON
set inav_gps_delay = 200
set inav_gps_min_sats = 6
set inav_w_z_baro_p =  0.350
set inav_w_z_gps_p =  0.200
set inav_w_z_gps_v =  0.200
set inav_w_xy_gps_p =  1.000
set inav_w_xy_gps_v =  2.000
set inav_w_z_res_v =  0.500
set inav_w_xy_res_v =  0.500
set inav_w_acc_bias =  0.010
set inav_max_eph_epv =  1000.000
set inav_baro_epv =  100.000
set nav_disarm_on_landing = ON
set nav_use_midthr_for_althold = OFF
set nav_extra_arming_safety = OFF
set nav_user_control_mode = ATTI
set nav_position_timeout = 5
set nav_wp_radius = 100
set nav_max_speed = 300
set nav_manual_speed = 500
set nav_manual_climb_rate = 200
set nav_landing_speed = 200
set nav_land_slowdown_minalt = 500
set nav_land_slowdown_maxalt = 2000
set nav_emerg_landing_speed = 500
set nav_min_rth_distance = 500
set nav_rth_tail_first = OFF
set nav_rth_alt_mode = AT_LEAST
set nav_rth_altitude = 1000
set nav_mc_bank_angle = 30
set nav_mc_hover_thr = 1500
set nav_mc_auto_disarm_delay = 2000
set nav_fw_cruise_thr = 1400
set nav_fw_min_thr = 1200
set nav_fw_max_thr = 1700
set nav_fw_bank_angle = 20
set nav_fw_climb_angle = 20
set nav_fw_dive_angle = 15
set nav_fw_pitch2thr = 10
set nav_fw_roll2pitch = 75
set nav_fw_loiter_radius = 5000
set serialrx_provider = SPEK1024
set spektrum_sat_bind = 0
set telemetry_switch = OFF
set telemetry_inversion = ON
set frsky_default_lattitude =  0.000
set frsky_default_longitude =  0.000
set frsky_coordinates_format = 0
set frsky_unit = IMPERIAL
set frsky_vfas_precision = 0
set hott_alarm_sound_interval = 5
set battery_capacity = 0
set vbat_scale = 110
set vbat_max_cell_voltage = 43
set vbat_min_cell_voltage = 33
set vbat_warning_cell_voltage = 35
set current_meter_scale = 400
set current_meter_offset = 0
set multiwii_current_meter_output = OFF
set current_meter_type = ADC
set align_gyro = DEFAULT
set align_acc = DEFAULT
set align_mag = DEFAULT
set align_board_roll = -10
set align_board_pitch = 30
set align_board_yaw = 0
set gyro_lpf = 42HZ
set moron_threshold = 32
set imu_dcm_kp = 2500
set imu_dcm_ki = 50
set imu_dcm_kp_mag = 10000
set imu_dcm_ki_mag = 0
set pos_hold_deadband = 20
set alt_hold_deadband = 50
set yaw_motor_direction = 1
set yaw_jump_prevention_limit = 200
set tri_unarmed_servo = ON
set servo_lowpass_freq = 400
set servo_lowpass_enable = OFF
set failsafe_delay = 10
set failsafe_off_delay = 200
set failsafe_throttle = 1000
set failsafe_kill_switch = OFF
set failsafe_throttle_low_delay = 100
set failsafe_procedure = SET-THR
set rx_min_usec = 885
set rx_max_usec = 2115
set acc_hardware = 0
set baro_use_median_filter = ON
set baro_hardware = 0
set mag_hardware = 0
set blackbox_rate_num = 1
set blackbox_rate_denom = 1
set blackbox_device = SPIFLASH
set magzero_x = 327
set magzero_y = -443
set magzero_z = 207
set acczero_x = 62
set acczero_y = -14
set acczero_z = 68
set accgain_x = 4104
set accgain_y = 4083
set accgain_z = 4031

# rxfail

rxfail 0 a
rxfail 1 a
rxfail 2 a
rxfail 3 a
rxfail 4 h
rxfail 5 h
rxfail 6 h
rxfail 7 h
rxfail 8 h
rxfail 9 h
rxfail 10 h
rxfail 11 h
rxfail 12 h
rxfail 13 h
rxfail 14 h
rxfail 15 h
rxfail 16 h
rxfail 17 h

# dump profile

# profile

profile 0

# aux

aux 0 1 0 900 1300
aux 1 2 0 1300 1700
aux 2 3 1 1300 1700
aux 3 10 2 1300 1700
aux 4 9 2 1700 2100
aux 5 12 2 1700 2100
aux 6 14 1 1700 2100
aux 7 0 0 900 900
aux 8 0 0 900 900
aux 9 0 0 900 900
aux 10 0 0 900 900
aux 11 0 0 900 900
aux 12 0 0 900 900
aux 13 0 0 900 900
aux 14 0 0 900 900
aux 15 0 0 900 900
aux 16 0 0 900 900
aux 17 0 0 900 900
aux 18 0 0 900 900
aux 19 0 0 900 900

# adjrange

adjrange 0 0 0 900 900 0 0
adjrange 1 0 0 900 900 0 0
adjrange 2 0 0 900 900 0 0
adjrange 3 0 0 900 900 0 0
adjrange 4 0 0 900 900 0 0
adjrange 5 0 0 900 900 0 0
adjrange 6 0 0 900 900 0 0
adjrange 7 0 0 900 900 0 0
adjrange 8 0 0 900 900 0 0
adjrange 9 0 0 900 900 0 0
adjrange 10 0 0 900 900 0 0
adjrange 11 0 0 900 900 0 0

# rxrange

rxrange 0 1000 2000
rxrange 1 1000 2000
rxrange 2 1000 2000
rxrange 3 1000 2000

# servo

servo 0 1000 2000 1500 90 90 100 -1
servo 1 1000 2000 1500 90 90 100 -1
servo 2 1000 2000 1500 90 90 100 -1
servo 3 1000 2000 1500 90 90 100 -1
servo 4 1000 2000 1500 90 90 100 -1
servo 5 1000 2000 1500 90 90 100 -1
servo 6 1000 2000 1500 90 90 100 -1
servo 7 1000 2000 1500 90 90 100 -1

set nav_alt_p = 50
set nav_alt_i = 0
set nav_alt_d = 0
set nav_vel_p = 100
set nav_vel_i = 50
set nav_vel_d = 10
set nav_pos_p = 65
set nav_pos_i = 120
set nav_pos_d = 0
set nav_posr_p = 180
set nav_posr_i = 15
set nav_posr_d = 100
set nav_navr_p = 10
set nav_navr_i = 5
set nav_navr_d = 8
set deadband = 5
set yaw_deadband = 5
set throttle_tilt_comp_str = 0
set default_rate_profile = 0
set gimbal_mode = NORMAL
set mag_declination = 240
set mag_hold_rate_limit = 90
set p_pitch = 40
set i_pitch = 30
set d_pitch = 30
set p_roll = 35
set i_roll = 30
set d_roll = 30
set p_yaw = 70
set i_yaw = 45
set d_yaw = 0
set p_level = 20
set i_level = 15
set d_level = 75
set max_angle_inclination_rll = 300
set max_angle_inclination_pit = 300
set gyro_soft_lpf_hz = 60
set acc_soft_lpf_hz = 15
set dterm_lpf_hz = 40
set yaw_lpf_hz = 30
set yaw_p_limit = 300

# dump rates

# rateprofile

rateprofile 0

set rc_expo = 70
set rc_yaw_expo = 20
set thr_mid = 50
set thr_expo = 0
set roll_rate = 50
set pitch_rate = 50
set yaw_rate = 40
set tpa_rate = 0
set tpa_breakpoint = 1500

# 

[blackbox_log_2016-07-17_205405.TXT](https://github.com/iNavFlight/inav/files/368093/blackbox_log_2016-07-17_205405.TXT)

**Comment by skaman82** [comment]
@Ralf-W setting gyro_sync_denom=12 does'n changed my looptime at all. It was still 2000. I didn't touched gyro_sync. It is OFF on default

**Comment by theArchLadder** [comment]
@skaman82 `gyro_sync` should be ON if you want to use `gyro_sync_denom`. Not sure why ON is not the default, it seems to be the better option...?

**Comment by Ralf-W** [comment]
If I remember correctly, for INAV looptime was set to a fixed value of 2000 for F1 processors to obtain a predictable behaviour even if users activate a lot of features in parallel (like LED, Display, etc.).  At that time, gyro_sync was set to OFF as default. However, with this issue here, discussion and experiments started to disable the hardware lpf and use software filtering to a higher extend. Doing this, it is a must to use gyro_sync = ON with a gyro_sync_denom that the FC can handle. For my Naze32 I can use gyro_sync_denom = 12 resulting in a looptime of 1500. It was tested that the minimum looptime for the F1 processor is about 1400. Boards with F2 (or F4) processors can handle much shorter looptimes and thus smaller gyro_sync_denom.

Edit:
@skaman82 
In your dump I find "set gyro_lpf = 42HZ" instead of "set gyro_lpf = 256HZ". With hardware filter set to 42HZ plus the softfilters discussed here, your copter will indeed become difficult to handle.

**Comment by WaspFPV** [comment]
How about changing defaults to:
set gyro_lpf = 188hz
set gyro_soft_lpf = 60
set gyro_sync = ON
set gyro_sync_denom = 2 (F1) or 1 (F3)

That should fly pretty much like cleanflight out of the box. The default of 42hz has too much latency imo.
![mpu_latency](https://cloud.githubusercontent.com/assets/4846411/16910332/659819a4-4cd8-11e6-9461-88c50f163bf5.jpg)
(thanks to @borisbstyle for the image)

**Comment by theArchLadder** [comment]
@andre92 I did notice a big difference when going from 188hz to 256hz (i think 256hz actually just disables the hardware filter).

**Comment by WaspFPV** [comment]
@theArchLadder 256hz in Inav is the same setting as OFF in CF/BF and it still filters at 256hz. The only setting that completely disables the filter is "Experimental" in betaflight but Boris doesn't recommend it, and BF uses 256hz by default. I think it would be too radical for a default on Inav, especially since current default is 42hz. Also, Inav doesn't run below 1khz, although it is nice to run 1500 looptime with 256hz on an F1 if you use ppm receiver and gps on uart2. (that is pretty much the limit of an F1, I had to enable emf_avoidance to get it to fly)

**Comment by theArchLadder** [comment]
@andre92 Using 256hz and relying on the software filters instead seems to give better performance. Perhaps it has something to do with the gyro running at 8khz...

In my testing with switching from 188hz to 256hz i could increase P-gain from 6.6 to 8.4 before constant oscillations, that's a huge difference, and the software filters still does a good job of filtering the noise.

Would be interesting to hear from more users trying 256hz and tuning the software filters on bigger copters... Anyone willing to try? @stronnag perhaps? @DzikuVx did you only test with 188hz?

**Comment by DzikuVx** [comment]
@theArchLadder @andre92 I'm flying 

```
gyro_lpf = 256HZ
gyro_sync=ON
gyro_sync_denom=12
gyro_soft_lpf_hz = 75
dterm_lpf_hz = 50
```

on SPRacingF3. Super nice experience

**Comment by theArchLadder** [comment]
@DzikuVx Is that on your 250? Have you tried it on anything bigger?

**Comment by stronnag** [comment]
@DzikuVx 

```
gyro_sync_denom=12
```

Isn't that loop time = 12000 ?

**Comment by WaspFPV** [comment]
@stronnag with gyro_lpf at 256hz, gyro runs at 8khz, with 125us interval instead of 1ms, so 12*125 will be 1500
@DzikuVx can an SPR F3 run Inav at 1khz? just curious. I make 1500 looptime on a Naze32 (F1). What is your cpu load with that looptime?

**Comment by stronnag** [comment]
gotcha. How confusing.

**Comment by theArchLadder** [comment]
@andre92 I run my F3 at 1khz. I think the minimum looptime is something like 600-700us

**Comment by DzikuVx** [comment]
@andre92 I was unable to use 1kHz on F3 due to some ESC glitch on OneShot125 (Emax ESC). Motors were spinning on boot, so back 1.5kHz.

@theArchLadder on 600mm quad I use 188HZ and looptime=2000

**Comment by theArchLadder** [comment]
@DzikuVx was 256hz bad on your 600mm or have you just not tried it yet?

**Comment by DzikuVx** [comment]
Did not tried yet. And probably will not try soon since my FC got burned there.

**Comment by Ralf-W** [comment]
I am experimenting with my "large" tricopter (Distance front motors to tail motor: 93 cm; 766 gr) and “set gyro_lpf = 256HZ” with a Naze32 (F1). Looptime is 1500. At the moment I have set the soft_lpf_hz to very low values. 
But for me, it performs excellently in smooth flight and in smooth acrobatics.

Here are two Flips. First graph is gyro(roll) and rc_command(roll). Second graph is pid_sum(roll) plus P,I,D (roll).
![flip](https://cloud.githubusercontent.com/assets/16436064/16943642/006d7130-4d9d-11e6-9854-55f75dd1df1e.jpg)
- Are my soft_lpf_hz set too low?
- Gyro follows rc_command at the beginning of the flip very fine. But the flip ends some (short) time after rc_command returned to center stick. And in the second flip the tricopter even overshoots the flip a bit. What should I change to solve these minor problems?

INAV/NAZE 1.1.0 May 11 2016 / 02:37:17 (1f49891)
Persistent config flags: 0x00000001
mixer TRI
set looptime = 2000
set emf_avoidance = OFF
set i2c_overclock = OFF
set gyro_sync = ON
set gyro_sync_denom = 12
set rc_smoothing = ON
set input_filtering_mode = OFF
set small_angle = 25
set gyro_lpf = 256HZ
set deadband = 5
set yaw_deadband = 5
set p_pitch = 50
set i_pitch = 20
set d_pitch = 180
set p_roll = 50
set i_roll = 30
set d_roll = 180
set p_yaw = 120
set i_yaw = 90
set d_yaw = 1
set p_level = 120
set i_level = 15
set d_level = 75
set max_angle_inclination_rll = 300
set max_angle_inclination_pit = 300
set gyro_soft_lpf_hz = 30
set acc_soft_lpf_hz = 15
set dterm_lpf_hz = 15
set yaw_lpf_hz = 30
set yaw_p_limit = 300

**Comment by DzikuVx** [comment]
1. I would say you have too much gyro noise, even with low lpf PID output looks very noisy. Balance props
2. 30Hz is very low. 
3. Remember about inertia. motors far from COG have huge inertia when they start to rotate and it's hard to stop them. Delay between command and execution (roll stop) is normal

**Comment by WaspFPV** [comment]
I suspect the overshoot is a result of overfiltering, as @DzikuVx points out. I have the same issues with my big noisy quadcopter. 
I wouldn't set gyro_soft_lpf_hz below 50 even with gyro_lpf = 256hz and a naze rev6, instead I'd try soft mounting the flight controller or troubleshooting the vibrations.

**Comment by stronnag** [comment]
Just for fun I tried a couple of things on my 420mm quad and 600mm tri (both previously on defaults of gyro_lpf 98Hz, gyro_sync=2)

On the quad, first used @DzikuVx's 256HZ / sync=12 et al settings. Flew manual, PH, RTH, WPs. Didn't really feel much improvement, and the downside was that landing detection was no longer instant (like taking 3-5 seconds on 500ms landing shutdown delay). 

I then when to 188HZ / sync=1 (1000 loop). Again manual,  PH, RTH, WPs. To my somewhat tame flying style, I liked it. And landing detection is good again. This is now my personal default on F3. Worked well on the 600mm tri as well, again in all flight modes.

**Comment by theArchLadder** [comment]
@stronnag That landing detection behavior sounds strange, could you share a bblog on that?

**Comment by stronnag** [comment]
Unfortunately not, the 256HZ log was largely corrupt. Very annoying. However it was consistent over two attempts.

**Comment by WaspFPV** [comment]
@stronnag, if you can use sync_denom = 1 with 188hz on a F3, you should use gyro_sync_denom = 8 with gyro_lpf = 256hz
you should be able to keep 1000 looptime on F3.
gyro_sync_denom = 12 is for F1 controllers. (or =2 when using 188hz or lower)

**Comment by theArchLadder** [comment]
@stronnag What values did you use for `gyro_soft_lpf_hz` and `dterm_lpf_hz`? I guess it makes sense to use lower values on a bigger copter (slower spinning props).

**Comment by DzikuVx** [comment]
@theArchLadder I also thought so, but well, I was wrong. As long as gyro noise is not making terrible things, it's worth to keep bot as high as possible. Just to find a sweet spot between delay and noise

**Comment by DzikuVx** [comment]
Everyone interested in this issue might want to try #369

**Comment by WaspFPV** [comment]
#369 looks good, but the motor output is still 666hz but with 2khz gyro sampling? I noticed in Betaflight the attitude estimation has been seperated from the pid loop (only 100hz and it still works), so task_pid is really only the acro controller. Is it possible to split the Inav pid task so attitude and position estimation are seperated and only the fp-pid function is treated as the "real" pid loop? That way maybe the gyro/pid and motor output can run 1khz with attitude estimation and position estimation every other loop.
Something like this: (should work even on F1 targets)
1khz: 
- acc sampling and filtering (betaflight does this at 1khz and the attitude loop takes the latest filter values)
- gyro sampling and filtering (maybe 2khz for F3)
- pid calculation and motor output

100-500hz:
- attitude estimation
- position estimation
- navigation updates

Since fp-pid has a smoothing filter on angleRate, I don't think it will fly jerky at lower angle updates.

**Comment by DzikuVx** [comment]
@andre92 OK, this deserves few longer sentences.
1. I'm working in a separate branch on something that I all "Race" feature. When enabled it will sacrifice as much as possible to give as much processing power for gyro/pid loop. One of the things will be acc running at much lower frequency (120Hz) but also with faster (less precise) ACC filtering. And maybe I will find some other things that can be disabled but will give some extra performance to gyros
2. Betaflight does ACC sampling at 100Hz
3. We can not do ACC at 100Hz in case of GPS assisted modes. Not enough resolution. That's why "Race" mode will have so much acc support as required to do emergency hover in Attitude or show artificial horizon on OSD
4. This might sound as heresy, but I do not belive updating motors above 500Hz makes much difference. OK, maybe 1kHz. Why? Because I did some math. Imagine a 5" prop running at 12000RPM and 2kHz motor update. And now, guess what is the distance prop tip travels in 500us (2kHz). It is only 40mm. more less 10% of a full rotation. How much power would be required to considerably change speed and thrust on a distance of 36 degrees? Even with Active Breaking.  Motors have limited power. Even ESC shows some problems here. By default most of them runs motor on 8kHz. That really does not give much room to change power delivered to motors in a smooth way.
5. But, BF running main pid loop very fast, at 2kHz or even 8kHz have some different advantages besides motor:
   1. faster gyro update what we are doing now in #369 
   2. Less delay
   3. better filtering
6. Running 8kHz gyro on I2C connected gyros is rather no-no. Bus not fast enough without DMA. And we do not have DMA ATM. In case of SPI it is better
7. I will try to get more, but gyro at 1kHz for F1 and 2.6kHz on I2C F3 is all we can do.

**Comment by WaspFPV** [comment]
I just checked acc sampling in Betaflight, it still runs faster than attitude loop, at 333hz on my Naze. My bad, I thought it was 1khz, as it should be. I think in Inav it should be at least 500hz as it is now.

Personally I don't need 2khz or more, but 1khz feels just a little more locked in than 500hz, especially in descents. For smaller copters this may be more noticable because of higher Nyquist frequencies.

I will do some testing and see what is possible on a Naze. I was able to achieve 1500us looptime, so I think there is room for one extra gyro + pid update in 500us. 

I think I'm going to break up the main pid loop into 3 tasks: gyro+pid, acc+attitude and positioning+navigation, and see how that works.

**Comment by borisbstyle** [comment]
@andre92 On F3 and F4 boads it is 1khz.....on F1 boards it is slower depending of your PID/gyro speed

**Comment by WaspFPV** [comment]
@borisbstyle I figured I can at least try to split up the pid loop into seperate tasks and run them all at a different rate. I like the idea of only handling acro controller in main pid loop and seperate attitude and navigation as you did in Betaflight, I'll see how that works out in Inav. Probably they can all run at 1khz on F3 targets, but I would really like to have gyro and acro controller at 1khz on a F1. Just something to try for myself first.

**Comment by borisbstyle** [comment]
@andre92 
I run gyro fast just because I use lpf of 256hz. When you sample too slow you will get aliasing artifacts.

Typical motor noise is between 200 and 400hz. To get a good sampling quality you would need to sample  x10 to be able to see the noise properly and filter it. I think the minimum sampling rate is than 400*10=4k

**Comment by WaspFPV** [comment]
Ok here's what I did. I pulled acc, attitude, position and navigation updates out of the main loop and made them into seperate tasks. Their period is set to masterconfig.looptime, so I can play around with the looptime setting for their period. So far it seems to work with gyro/pid/motors at 1khz and looptime = 3000. (gyro_sync and emf_avoidance ON) I'll try to fly it tomorrow. I made a fork with my changes, test at your own risk.

**Comment by digitalentity** [comment]
Another suggestion - get rid from synchronous motor update and stick with unsynched, at 490Hz for PWM and, i.e. 2kHz for OneShot. As @DzikuVx mentioned, motors are not going to change thrust much for 500usec anyhow.

**Comment by digitalentity** [comment]
IMO sampling acc at least at 500Hz rate is a must for reducing effects of aliasing. Basically, we need 3 rates of sampling:
1) gyro
2) acc
3) PID
And as far as I'm concerned only gyro should be read in a synchronous manner, the rest can tolerate some jitter without noticable side-effects.

**Comment by digitalentity** [comment]
@borisbstyle your expertise on this matter is greatly appreciated - you have similar separation in Betaflight for ages now. :smile:

**Comment by borisbstyle** [comment]
@digitalentity 
I indeed agree about the motors side. It is only the delay what improves, but thats really marginal compared to real PID perforamance variables.
There are other variables too what affect the acro performance even more.

I didnt have much time to dig into your current state of the code, but I will soon once I finish 3.0 what needs some attention now.

**Comment by kronosol** [comment]
Hey, to improve the Acro flight performance,.. how about supporting OneShot42 or better Multishot?
[Multishot Raceflight FW](http://www.rcgroups.com/forums/showthread.php?t=2540753)
Edit: [Multishot & 4k4k Mode](https://flyingbearfpv.com/2016/04/19/whats-this-4k4k-mode-and-multishot/)

**Comment by digitalentity** [comment]
@kronosol 
#397 will open possibility to implement OneShot42 or MultiShot. Unsynched mode may introduce slight delay in transitioning command to motors, however it's really marginal as @DzikuVx pointed out that motor thrust doesn't change that quickly anyway.

**Comment by DzikuVx** [comment]
@digitalentity @kronosol probably delay from unsynched motor update will be much much lower than gain of at least 3 times faster update of OneShot42 comparing to OneShot125.

**Comment by digitalentity** [comment]
@DzikuVx update rate is defined by `motor_pwm_rate`, OneShot125 can be dialed as fast as almost 8kHz, why would you need to go faster than that? I might be wrong, but I really don't see a benefit of using OneShot42 or Multishot over OneShot125.

**Comment by DzikuVx** [comment]
@digitalentity yes, OneShot125 can be executed at ~8kHz. Still, it takes up to 125us to push data to motor. With OneShot42 it takes up to 42us. 80us less delay. How much delay async motor update can introduce?

**Comment by digitalentity** [comment]
@DzikuVx in theory as much as `motor_pwm_rate` period (half of that on average). If OneShot42 can be updated faster than 8kHz - we need it, if not - drop it and stick with OneShot125 as it's more noise-tolerant.
EDIT: This delay is totally independent of used protocol - OS42 and OS125 both will have ~62us delay on average if run at 8kHz rate.

**Comment by digitalentity** [comment]
@DzikuVx based on your math even 2kHz rate (500us delay) doesn't matter much, how much 62us would hurt in terms of flight performance?
Also bear in mind that MPU6500 alone running with gyro_lpf=256HZ has an inherent delay of 970us.

**Comment by DzikuVx** [comment]
@digitalentity now I'm referring to smth completely different. I'm not talking about changing motor speed at 2kHz. I'm talking about how long it takes for a motor to start reacting to new value. 

With PWM at 490Hz it takes ~2000us to notify ESC about a change. This is only delay introduced by protocol. So, with PID loop at 500Hz and up to 2000us of protocol delay, ESC can _start_ to react 4ms (worst case)  _after_ gyro got a change (2000us waiting for next PID loop and 2000us to transfer data to ESC. 

With OneShot125 and looptime 2000 it is only 2150us (assuming motors were synced right after PID loop). With OneShot42 is is only 2042us and so on.

**Comment by digitalentity** [comment]
Ok, I  think this is solved now

**Comment by alexrjohnston** [comment]
Hi guys, sorry to grave dig. Just a quick note that the code below transformed my 330 with 8" dji props from a wobbly wreck to a locked in beast of a machine on inav 1.3:

set gyro_sync_denom = 12
set gyro_lpf = 256hz
set gyro_sync = ON
set gyro_soft_lpf_hz = 75
looptime 1k on sp racing f3

We also have a 650 hex im going to try the same settings with once we have rebuilt the air frame. Before making the above changes both machines had a slow oscillation during hover even in manual mode that I could never seem to tune out without backing everything off too far. Big thanks to everyone who has contributed to this awesome project.

**Comment by digitalentity** [comment]
@alexrjohnston you are welcome! Indeed high values of `gyro_lpf` make a big difference. I wonder why PixHawk devs think otherwise...

**Comment by DzikuVx** [comment]
@alexrjohnston I'm very glad to hear I liked how your quad behaved. Only bear in mind, that if you have bigger propellers, gyro noise frequency will go down and you might want to lower `gyro_soft_lpf_hz`. My 600mm quad on 12" props required `gyro_soft_lpf_hz` of < 60

**Comment by alexrjohnston** [comment]
great info, thanks for the advice! I also tried gyro_sync_denom = 8  yesterday on the f3 with pos and alt hold and it flew even nicer. is there any danger doing this? the cpu seemed fine until gyro_sync_denom = 4 which it couldn't handle. - Al.

**Comment by DzikuVx** [comment]
Generally, the lower the better. As long as CPU can hold the looptime with GPS and stuff, it is fine

**Comment by alexrjohnston** [comment]
ideal, thanks again! seriously impressed with inav since making these adjustments.
